src/Iterators.jl

```diff
@@ -311,16 +311,13 @@ if VERSION >= v"0.4-dev"
 else
     eltype(p::Product) = tuple(map(eltype, p.xss)...)
 end
-length(p::Product) = prod(map(length, p.xss))
+length(p::Product) = isempty(p.xss) ? 1 : prod(map(length, p.xss))
```

Comment:
Great! This is much faster, probably because it avoids a type unstable array allocation. :100: